### PR TITLE
Simplify support for Guava Optional<T> in dropwizard-jersey

### PR DESCRIPTION
**Pull request by joschi**
Instead of multiple providers for each type of JAX-RS parameter (e. g. @QueryParam or @HeaderParam) we now have a single ParamConverterProvider.

Refs jersey/jersey#98, [JERSEY-2612](https://java.net/jira/browse/JERSEY-2612)

### CODE REVIEW

File: dropwizard-jersey/src/main/java/io/dropwizard/jersey/guava/OptionalParamConverterProvider.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+                @Override
+                public String toString(final T value) throws IllegalArgumentException {
+                    return value.toString();
+                }
+            };
+        }
+        final Set<ParamConverterProvider> converterProviders = Providers.getProviders(locator, ParamConverterProvider.class);
+        for (ParamConverterProvider provider : converterProviders) {
+            @SuppressWarnings("unchecked")
```
Comment (Trundle) [inline dropwizard-jersey/src/main/java/io/dropwizard/jersey/guava/OptionalParamConverterProvider.java]: Does the `@SuppressWarnings` here have any effect? Can't really see what could cause a unchecked warning right now.

File: dropwizard-jersey/src/main/java/io/dropwizard/jersey/guava/OptionalParamConverterProvider.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+    private final ServiceLocator locator;
+
+    @Inject
+    public OptionalParamConverterProvider(final ServiceLocator locator) {
+        this.locator = locator;
+    }
+
+    @Override
+    public <T> ParamConverter<T> getConverter(final Class<T> rawType, final Type genericType, final Annotation[] annotations) {
```
Comment (Trundle) [inline dropwizard-jersey/src/main/java/io/dropwizard/jersey/guava/OptionalParamConverterProvider.java]: I think there should be a check whether it's really an `Optional` (e.g. `if (!Optional.class.equals(rawType)) { return null; }`). Right now, a converter is returned for any generic that has a type parameter for which some converter exists, resulting in a `ClassCastException` when being used (lines 37 and 54).

File: dropwizard-jersey/src/main/java/io/dropwizard/jersey/guava/OptionalParamConverterProvider.java (hunk 1 above)
Comment (joschi) [inline dropwizard-jersey/src/main/java/io/dropwizard/jersey/guava/OptionalParamConverterProvider.java]: You're completely correct. The suppression of unchecked warnings is not necessary here.

File: dropwizard-jersey/src/main/java/io/dropwizard/jersey/guava/OptionalParamConverterProvider.java (hunk 2 above)
Comment (joschi) [inline dropwizard-jersey/src/main/java/io/dropwizard/jersey/guava/OptionalParamConverterProvider.java]: Whoops. Of course the actual type of the parameter should be checked. Thanks!

**Comment by Trundle** [comment]
LGTM

**Comment by carlo-rtr** [comment]
Nice!

**Pull request merged**